cmd/coordinated/metrics.go

```diff
@@ -4,38 +4,75 @@
 package main
 
 import (
+	"context"
+	"math"
+	"time"
+
 	"github.com/diffeo/go-coordinate/coordinate"
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
 )
 
-var coordinateSummary = prometheus.NewGaugeVec(
-	prometheus.GaugeOpts{
-		Namespace: "diffeo",
-		Subsystem: "coordinate",
-		Name:      "coordinate_summary",
-		Help:      "Summary of coordinate work specs",
-	},
-	[]string{
-		"namespace",
-		"work_spec",
-		"status",
-	},
+var (
+	workUnitsNumber = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Namespace: "coordinate",
+			Name:      "summary_seconds",
+			Help:      "Seconds required to gather coordinate summary",
+			Buckets:   prometheus.ExponentialBuckets(math.Pow(2, -5), 2, 12),
+		})
+
+	summarySeconds = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "coordinate",
+			Name:      "work_units",
+			Help:      "Number of coordinate work specs",
+		},
+		[]string{
+			"namespace",
+			"work_spec",
+			"status",
+		})
 )
 
 func init() {
-	prometheus.MustRegister(coordinateSummary)
+	prometheus.MustRegister(summarySeconds)
+	prometheus.MustRegister(workUnitsNumber)
 }
 
-func observe(coord coordinate.Coordinate) {
+// Observe repeatedly calls Summarize() on coordinate in an infinite loop, and
+// observes each SummaryRecord's fields on a prometheus GaugeVec, and the
+// resultant time duration on a prometheus Histogram.
+func Observe(
+	ctx context.Context,
+	coord coordinate.Coordinate,
+	period time.Duration,
+	log *logrus.Logger,
+) {
 	for {
-		summary, _ := coord.Summarize()
-		for _, record := range summary {
-			statusBytes, _ := record.Status.MarshalJSON()
-			coordinateSummary.With(prometheus.Labels{
-				"namespace": record.Namespace,
-				"work_spec": record.WorkSpec,
-				"status":    string(statusBytes),
-			}).Set(float64(record.Count))
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(period):
```

Comment:
This is fine here for now.

If you write this loop again, note that this creates a new timer and a new channel every time through the loop.  (In this specific case, either the timer has triggered or the context done channel has closed and you're exiting so it's okay.)